compiler/noirc_frontend/src/signed_field.rs

```diff
@@ -0,0 +1,160 @@
+pub struct SignedField {
+    pub field: FieldElement,
+    pub is_negative: bool,
+}
+
+impl SignedField {
+    pub fn new(field: FieldElement, is_negative: bool) -> Self {
+        Self { field, is_negative }
+    }
+
+    pub fn positive(field: impl Into<FieldElement>) -> Self {
+        Self { field: field.into(), is_negative: false }
+    }
+
+    /// Convert a signed integer to a SignedField, carefully handling
+    /// INT_MIN in the process. Note that to convert an unsigned integer
+    /// you can call `SignedField::positive`.
+    #[inline]
+    pub fn from_signed<T>(value: T) -> Self
+    where
+        T: num_traits::Signed + AbsU128,
+    {
+        let negative = value.is_negative();
+        let value = value.abs_u128();
+        SignedField::new(value.into(), negative)
+    }
+
+    /// Convert a SignedField into an unsigned integer type (up to u128),
+    /// returning None if the value does not fit (e.g. if it is negative).
+    #[inline]
+    pub fn try_to_unsigned<T: TryFrom<u128>>(self) -> Option<T> {
+        if self.is_negative {
+            return None;
+        }
+
+        assert!(std::mem::size_of::<T>() <= std::mem::size_of::<u128>());
+        let u128_value = self.field.try_into_u128()?;
+        u128_value.try_into().ok()
+    }
+
+    /// Convert a SignedField into a signed integer type (up to i128),
+    /// returning None if the value does not fit. This function is more complex
+    /// for handling negative values, specifically INT_MIN which we can't cast from
+    /// a u128 to i128 without wrapping it.
+    #[inline]
+    pub fn try_to_signed<T>(self) -> Option<T>
+    where
+        T: TryFrom<u128> + TryFrom<i128> + num_traits::Signed + num_traits::Bounded + AbsU128,
+        u128: TryFrom<T>,
+    {
+        let u128_value = self.field.try_into_u128()?;
+
+        if self.is_negative {
+            // The positive version of the minimum value of this type.
+            // E.g. 128 for i8.
+            let positive_min = T::min_value().abs_u128();
+
+            // If it is the min value, we can't negate it without overflowing
+            // so test for it and return it directly
+            if u128_value == positive_min {
+                Some(T::min_value())
+            } else {
+                let i128_value = -(u128_value as i128);
+                T::try_from(i128_value).ok()
+            }
+        } else {
+            T::try_from(u128_value).ok()
+        }
+    }
+}
+
+impl std::ops::Neg for SignedField {
+    type Output = Self;
+
+    fn neg(mut self) -> Self::Output {
+        self.is_negative = !self.is_negative;
+        self
+    }
+}
+
+impl Ord for SignedField {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.is_negative != other.is_negative {
+            if self.is_negative {
+                std::cmp::Ordering::Less
+            } else {
+                std::cmp::Ordering::Greater
+            }
+        } else {
+            self.field.cmp(&other.field)
```

Comment:
If both are negative values should the comparison be the other way around? Otherwise I think `-1` will be less than `-2`.

Reply:
(if this was the case, it might be good to add a small test for this)

Reply:
Great catch! Fixed